run0/src/container/mod.rs

```diff
@@ -0,0 +1,107 @@
+use std::path::PathBuf;
+
+use oci_spec::runtime::Spec;
+
+use crate::container::environment::Environment;
+use mounts::Mounts;
+use namespaces::Namespaces;
+
+mod environment;
+mod mounts;
+mod namespaces;
+
+/// Some OCI constants useful for our container implementation.
+const OCI_RUNTIME_SPEC_FILE: &str = "config.json";
+const OCI_RUNTIME_SPEC_ROOTFS: &str = "rootfs";
+
+/// Containers related errors
+#[derive(Debug)]
+pub enum Error {
+    OCISpecificationLoad(oci_spec::OciSpecError),
+    OCIInvalidNamespace(oci_spec::runtime::LinuxNamespaceType),
+    ContainerSpawnCommand(unshare::Error),
+    ContainerWaitCommand(std::io::Error),
+    ContainerExit(i32),
+    Unmount(std::io::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The `Container` struct provides a simple way to
+/// create and run a container on the host.
+pub struct Container {
+    /// The path to the OCI bundle used by the container
+    bundle: PathBuf,
+    /// The path to the rootfs used by the container
+    rootfs: PathBuf,
+    /// The namespaces which will be given to the container
+    namespaces: Namespaces,
+    /// The additional mounts mounted into the container beyond `rootfs`.
+    mounts: Mounts,
+    /// The container environment
+    environment: Environment,
+}
+
+impl Container {
+    /// Build a new container with the bundle provided in parameters.
+    pub fn new(bundle_path: &str) -> Result<Self> {
+        let bundle = PathBuf::from(bundle_path);
+
+        // Load the specification from the file
+        let spec =
+            Spec::load(&bundle.join(OCI_RUNTIME_SPEC_FILE)).map_err(Error::OCISpecificationLoad)?;
+
+        // Get the container rootfs from the OCI specification, and if not present, set to
+        // the default `bundle_path/rootfs`
+        let rootfs = spec
+            .root()
+            .as_ref()
+            .map_or(bundle.clone().join(OCI_RUNTIME_SPEC_ROOTFS), |x| {
+                bundle.clone().join(x.path().to_path_buf())
+            });
+
+        // Get the container namespaces if the linux block is defined into the specification.
+        let namespaces = if let Some(linux) = spec.linux() {
+            Namespaces::from(linux.namespaces())
+        } else {
+            Namespaces::default()
+        };
+
+        let environment = Environment::from(spec.process());
+
+        Ok(Container {
+            bundle: bundle.clone(),
+            rootfs: rootfs.clone(),
+            mounts: Mounts::default(),
```

Comment:
You can skip that one and use `..Default::default()` at the end of the `Container` initialization:

```rust
Ok(Container {
             bundle: bundle.clone(),
             rootfs: rootfs.clone(),
             environment: Environment::from(spec.process()),
             namespaces,
             ..Default::default()
}            
```